python/src/lib.rs

```diff
@@ -396,6 +397,185 @@ impl RawDeltaTable {
         Ok(serde_json::to_string(&metrics).unwrap())
     }
 
+    #[allow(clippy::too_many_arguments)]
+    #[pyo3(signature = (source,
+        predicate,
+        source_alias = None,
+        target_alias = None,
+        safe_cast = false,
+        writer_properties = None,
+        matched_update_updates = None,
+        matched_update_predicate = None,
+        // matched_update_all,
+        matched_delete_predicate = None,
+        matched_delete_all = None,
+        not_matched_insert_updates = None,
+        not_matched_insert_predicate = None,
+        // not_matched_insert_all,
+        not_matched_by_source_update_updates = None,
+        not_matched_by_source_update_predicate = None,
+        not_matched_by_source_delete_predicate = None,
+        not_matched_by_source_delete_all = None,
+    ))]
+    pub fn merge_execute(
+        &mut self,
+        source: PyArrowType<RecordBatch>,
+        predicate: String,
+        source_alias: Option<String>,
+        target_alias: Option<String>,
+        safe_cast: bool,
+        writer_properties: Option<HashMap<String, usize>>,
+        matched_update_updates: Option<HashMap<String, String>>,
+        matched_update_predicate: Option<String>,
+        // matched_update_all: Option<bool>,
+        matched_delete_predicate: Option<String>,
+        matched_delete_all: Option<bool>,
+        not_matched_insert_updates: Option<HashMap<String, String>>,
+        not_matched_insert_predicate: Option<String>,
+        // not_matched_insert_all: Option<bool>,
+        not_matched_by_source_update_updates: Option<HashMap<String, String>>,
+        not_matched_by_source_update_predicate: Option<String>,
+        not_matched_by_source_delete_predicate: Option<String>,
+        not_matched_by_source_delete_all: Option<bool>,
+    ) -> PyResult<String> {
+        let ctx = SessionContext::new();
+        let source_df = ctx.read_batch(source.0).unwrap();
```

Comment:
Instead of `read_batch`, create a [MemTable](https://docs.rs/datafusion/latest/datafusion/datasource/memory/struct.MemTable.html) from the batches in the reader.